src/relay/dnsrelay/upstream.rs

```diff
@@ -0,0 +1,59 @@
+use std::io::Result;
+use std::net::{IpAddr, Ipv4Addr, Ipv6Addr, SocketAddr, UdpSocket};
+use trust_dns_proto::op::Message;
+
+pub trait Upstream {
+    async fn lookup(&self, request: &Message) -> Result<Message>;
```

Comment:
Why your `trait` can have `async fn`? [Shock]

Reply:
I don't know a lot about rust and this is a WIP. At least we can do this in Kotlin. 🤣 

So I guess it should be `fn lookup(&self, request: &Message) -> Future<Result<Message>>`?

Reply:
Use a helper macro: `#[async_trait]`